static/js/components/Editor.js

```diff
@@ -0,0 +1,310 @@
+    this.formHelpers = {
+      getForm,
+      ...actionCreators
+    }
+
+    const { initialValue } = props
+    const editorStateOptions = {
+      doc:     defaultMarkdownParser.parse(initialValue || ""),
+      plugins: exampleSetup({
+        schema,
+        menuBar: false
+      })
+    }
+
+    this.state = {
+      editorState:  EditorState.create(editorStateOptions),
+      showLinkMenu: false
+    }
+    this.dispatchTransaction = this.dispatchTransaction.bind(this)
+  }
+
+  componentDidMount() {
+    const { autoFocus } = this.props
+    const { editorState } = this.state
+
+    if (this.node.current) {
+      this.view = new EditorView(this.node.current, {
+        state:               editorState,
+        dispatchTransaction: this.dispatchTransaction
+      })
+
+      if (autoFocus) {
+        this.view.focus()
+      }
+    }
+  }
+
+  dispatchTransaction(transaction: Object) {
+    // we need our own handler for this so we can hook into
+    // the EditorState lifecycle to synchronize it with the
+    // redux store (via the onChange prop)
+    const { onChange } = this.props
+
+    const state = this.view.state.apply(transaction)
+    this.view.updateState(state)
+    this.setState({ editorState: state })
+
+    onChange(defaultMarkdownSerializer.serialize(this.view.state.doc))
+  }
+
+  menuClickHandlerCreator = (command: Function) => (e: Object) => {
+    e.preventDefault()
+    command(this.view.state, this.view.dispatch, this.view)
+    this.view.focus()
+  }
+
+  renderMenuBar = () => {
+    const selectEmpty = this.selectionIsEmpty()
+
+    const linkIcon =
+      this.view && this.markIsActive(schema.marks.link) ? "link_off" : "link"
+
+    return (
+      <div className="menu-bar">
+        {menuBarManifest.map((buttonGroup, idx) => (
+          <div className="button-group" key={idx}>
+            {buttonGroup.map(({ command, title, icon, mark }) => (
+              <div
+                className={`menu-button ${
+                  mark && this.markIsActive(mark) ? "active" : ""
+                }`}
+                onClick={this.menuClickHandlerCreator(command)}
+                key={title}
+              >
+                {icon}
+              </div>
+            ))}
+          </div>
+        ))}
+        <div className="button-group">
+          <div
+            className={`menu-button ${selectEmpty ? "disabled" : ""}`}
+            onClick={selectEmpty ? null : this.handleLinkClick}
+          >
+            <i className={`material-icons ${linkIcon}`}>{linkIcon}</i>
+          </div>
+        </div>
+      </div>
+    )
+  }
+
+  markIsActive = (mark: Object) => {
+    // on the first render call this.view isn't there yet
+    if (!this.view) {
+      return false
+    }
+
+    const { from, $from, to, empty } = this.view.state.selection
+    return empty
+      ? !!mark.isInSet(this.view.state.storedMarks || $from.marks())
+      : this.view.state.doc.rangeHasMark(from, to, mark)
```

Comment:
Can you include a brief comment(s) about this logic? Not clear what `empty` means and what exactly is being done as a result of that value